### PR TITLE
SSH protocol support

### LINKED ISSUE #2159: Invalid bind address protocol: ssh:// in DOCKER_HOST
Hi there,

Docker 18.09 now supports `docker -H ssh://me@server`

https://github.com/docker/cli/pull/1014

I tried using it with `docker-compose`, but got an error.
I found this error caused by `docker-py` instead of `docker-compose`.
So, I will report this here.

## version
docker-py version: 3.4.1

### docker client
* Mac OSX 10.14 (Mojave)
* Docker for Mac Edge (2.0.0.0-beta1-mac75 2018-09-14)
* Docker CE 18.09.0-ce-beta1
* docker-compose 1.22.0

```
$ docker version
Client: Docker Engine - Community
 Version:           18.09.0-ce-beta1
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        78a6bdb
 Built:             Thu Sep  6 22:41:53 2018
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0-ce-beta1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       78a6bdb
  Built:            Thu Sep  6 22:49:35 2018
  OS/Arch:          linux/amd64
  Experimental:     true

$ docker-compose version
docker-compose version 1.22.0, build f46880f
docker-py version: 3.4.1
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018
```

### docker daemon
* AWS EC2
* Ubuntu 18.04 LTS
* Docker CE 18.09.0-beta5

```
$ docker version
Client:
 Version:           18.09.0-beta5
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        e1910c5
 Built:             Tue Oct 23 22:25:31 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0-beta5
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       e1910c5
  Built:            Tue Oct 23 21:50:22 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

## Configuration
```
$ cat docker-compose.yml
version: '3'
services:
  web:
    image: nginx:alpine
    ports:
      - "80:80"
```

## Expected Behavior

`docker-py` should parse `ssh://` as a valid protocol,
and `docker-compose` command can connect to remote the docker daemon.

## Actual Behavior
Failed to parse protocol `ssh://`

```
$ docker-compose -H ssh://docker-18-09-test up -d
Traceback (most recent call last):
  File "docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 124, in perform_command
  File "compose/cli/command.py", line 41, in project_from_options
  File "compose/cli/command.py", line 121, in get_project
  File "compose/cli/command.py", line 92, in get_client
  File "compose/cli/docker_client.py", line 127, in docker_client
  File "site-packages/docker/api/client.py", line 118, in __init__
  File "site-packages/docker/utils/utils.py", line 256, in parse_host
docker.errors.DockerException: Invalid bind address protocol: ssh://docker-18-09-test
[8412] Failed to execute script docker-compose
```

```
$ DOCKER_HOST=ssh://docker-18-09-test docker-compose up -d
Traceback (most recent call last):
  File "docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 124, in perform_command
  File "compose/cli/command.py", line 41, in project_from_options
  File "compose/cli/command.py", line 121, in get_project
  File "compose/cli/command.py", line 92, in get_client
  File "compose/cli/docker_client.py", line 127, in docker_client
  File "site-packages/docker/api/client.py", line 118, in __init__
  File "site-packages/docker/utils/utils.py", line 256, in parse_host
docker.errors.DockerException: Invalid bind address protocol: ssh://docker-18-09-test
[8460] Failed to execute script docker-compose
```

It would  be great if `docker-py` supports `ssh://` in DOCKER_HOST
Thanks!

### PR DESCRIPTION
**Pull request by shin-**
Resolves #2159 

Includes a fix for #2082

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks a lot for working on this, could you also support specifying `username@` ?

https://github.com/docker/cli/blob/master/cli/connhelper/ssh/ssh_test.go

**Comment by shin-** [comment]
@AkihiroSuda That should work already: https://github.com/docker/docker-py/pull/2165/files#diff-2dd7e931899af21cd0ae5d3b517c0f9eR87

File: docker/transport/sshconn.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    def __init__(self, base_url, timeout=60,
+                 pool_connections=constants.DEFAULT_NUM_POOLS):
+        self.ssh_client = paramiko.SSHClient()
+        self.ssh_client.load_system_host_keys()
+
+        parsed = urllib.parse.urlparse(base_url)
+        self.ssh_client.connect(
+            parsed.hostname, parsed.port, parsed.username,
+        )
```
Comment (AkihiroSuda) [inline docker/transport/sshconn.py]: Can we refuse `parsed.password`, `parsed.path`, `parsed.fragment`, `parsed.query` explicitly?

**Comment by pmb0** [comment]
Hello, what do I have to do to be able to use this function? When running I get the exception `Install paramiko package to enable ssh:// support`. `pip install paramiko` does not help.

```sh
$ DOCKER_HOST=ssh://<some-host> docker-compose up -d
Traceback (most recent call last):
  File "site-packages/docker/api/client.py", line 151, in __init__
NameError: name 'SSHAdapter' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 124, in perform_command
  File "compose/cli/command.py", line 42, in project_from_options
  File "compose/cli/command.py", line 123, in get_project
  File "compose/cli/command.py", line 94, in get_client
  File "compose/cli/docker_client.py", line 127, in docker_client
  File "site-packages/docker/api/client.py", line 156, in __init__
docker.errors.DockerException: Install paramiko package to enable ssh:// support
[45766] Failed to execute script docker-compose
```

**Comment by shin-** [comment]
It's not supported in Compose yet. It will be in the next release (1.24) which should come out sometime in January.

**Comment by pmb0** [comment]
Ah, thank you!

**Comment by technoe** [comment]
Still broken. Traceback (most recent call last):
  File "site-packages/docker/api/client.py", line 151, in __init__
NameError: name 'SSHAdapter' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 124, in perform_command
  File "compose/cli/command.py", line 42, in project_from_options
  File "compose/cli/command.py", line 123, in get_project
  File "compose/cli/command.py", line 94, in get_client
  File "compose/cli/docker_client.py", line 127, in docker_client
  File "site-packages/docker/api/client.py", line 156, in __init__
docker.errors.DockerException: Install paramiko package to enable ssh:// support
[23282] Failed to execute script docker-compose

**Comment by shin-** [comment]
> `docker.errors.DockerException: Install paramiko package to enable ssh:// support`

Damn, sometimes tracebacks give useful information, imagine that.

**Comment by technoe** [comment]
If only it were that easy. Paramiko is installed on the server and client.

**Comment by shin-** [comment]
The best thing to do in that case is to create a proper bug report with all the information needed for us to reproduce the issue. :)

**Pull request merged**